### PR TITLE
[CARBONDATA-3119] Fixed SDK Write for Complex Array Type when Array is Empty

### PR DESCRIPTION
**Pull request by shivamasn**
### What was the issue?
For SDK Write , it was going into bad record by returning null on passing empty array for Complex Type.
### What has been changed?
Added a check for empty array. This will return an empty array.

 - [ ] Any interfaces changed?
 NA
 - [ ] Any backward compatibility impacted?
 NA
 - [ ] Document update required?
NA
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
Test  case added
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala [hunk 1]
```diff
@@ -39,6 +39,27 @@ class TestComplexDataType extends QueryTest with BeforeAndAfterAll {
       .addProperty(CarbonCommonConstants.CARBON_BAD_RECORDS_ACTION, badRecordAction)
   }
 
+  test("test Projection PushDown for Array - String type when Array is Empty") {
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_BAD_RECORDS_ACTION, "FAIL")
```
Comment (ajantha-bhat) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala]: Take a previous CARBON_BAD_RECORDS_ACTION property as backup and set it back after this test case, else it affects other test suites

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala [hunk 2]
```diff
@@ -39,6 +39,27 @@ class TestComplexDataType extends QueryTest with BeforeAndAfterAll {
...
+    sql("drop table if exists table1")
+    sql("create table table1 (detail array<string>) stored by 'carbondata'")
+    sql("insert into table1 values('')")
+    checkAnswer(sql("select detail[0] from table1"), Seq(Row("")))
+    sql("drop table if exists table1")
+  }
+
+  test("test Projection PushDown for Struct - Array type when Array is Empty") {
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_BAD_RECORDS_ACTION, "FAIL")
```
Comment (ajantha-bhat) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala]: same as above

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 3]
```diff
@@ -23,8 +23,11 @@
 import java.util.*;
 
 import org.apache.avro.generic.GenericData;
+
```
Comment (ajantha-bhat) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: please revert unwanted changes, only your changes should be there in diff

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 4]
```diff
@@ -2072,4 +1797,35 @@ public void testReadingNullValues() {
     }
   }
 
+  @Test public void testSdkWriteWhenArrayOfStringIsEmpty()
+      throws IOException, InvalidLoadOptionException {
+
+    CarbonProperties.getInstance()
+        .addProperty(CarbonCommonConstants.CARBON_BAD_RECORDS_ACTION, "FAIL");
```
Comment (ajantha-bhat) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: same as above, take a backup of CARBON_BAD_RECORDS_ACTION and set it at the end of the test case

**Comment by ajantha-bhat** [comment]
@ravipesala , @chenliang613 : please add @shivamasn to whitelist

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by ravipesala** [comment]
add to whitelist

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1720/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1932/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9980/

**Comment by shivamasn** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1933/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9981/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1730/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1943/

**Comment by shivamasn** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1739/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9998/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1950/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10019/

**Comment by shivamasn** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1978/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1767/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10025/

**Comment by ajantha-bhat** [comment]
LGTM

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**